core/modules/filter/filter.pages.inc

```diff
@@ -491,28 +491,159 @@ function filter_format_editor_dialog_save($form, &$form_state) {
+ * disabled.
+ *
+ * @param string $url
+ *
+ * @return string
+ *   A cleaned-up URL that is either a properly formatted external URL or an
+ *   internal path.
+ *
+ * @see _filter_format_editor_link_url_validate()
+ */
+function _filter_cleanup_url($url) {
+  // Will add a default protocol if missing on an external url.
+  $url = _filter_create_valid_external_url($url);
+
+  // If it's an external URL with a protocol we can skip the rest of the
+  // clean up.
+  if (_filter_url_has_protocol($url)) {
+    return $url;
+  }
+
+  // Return anchors as is.
+  if (substr($url, 0, 1) == '#') {
+    return $url;
+  }
+
+  // Break up the components so we can build up an aliased internal path.
+  $url_parts = backdrop_parse_url($url);
+
+  // Lookup valid internal paths if multilingual
+  if (module_exists('locale') && language_multilingual()) {
+    if ($language = _filter_lookup_locale_path($url)) {
+      $url_parts['language'] = $language;
+    }
+  }
+
+  return url($url_parts['path'], $url_parts);
+}
+
+/**
+ * Forms a valid external URL if possible.
+ *
+ * Trims whitespace and automatically adds an http:// to addresses that
+ * appear to be external URLs but have no specified protocol.
+ *
+ * @param string $url
+ *   The URL entered by the user.
+ *
+ * @return string
+ *   The URL with the protocol prepended if needed.
+ *
+ * @see link_cleanup_url().
+ */
+function _filter_create_valid_external_url($url) {
+  // Skip if it's already a valid external URL with a protocol.
+  if (_filter_url_has_protocol($url)) {
+    return $url;
+  }
+
+  // If it looks like a domain name but it's missing a protocol Add an
+  // an http:// or https://, depending on the site configuration.
+  $domain_match = preg_match('/^(([a-z0-9]([a-z0-9\-_]*\.){1,63})([a-z]{2,63}))/i', $url);
+  if (!empty($domain_match)) {
+    if (isset($_SERVER['HTTPS']) &&
+    ($_SERVER['HTTPS'] == 'on' || $_SERVER['HTTPS'] == 1) ||
+    isset($_SERVER['HTTP_X_FORWARDED_PROTO']) &&
+    $_SERVER['HTTP_X_FORWARDED_PROTO'] == 'https') {
+      $protocol = 'https://';
+    }
+    else {
+      $protocol = 'http://';
+    }
+    $url = $protocol . $url;
+  }
+
+  return $url;
+}
+
+/**
+ * Checks if a URL has a protocol.
+ *
+ * Checks for a protocol as a sign that the URL is external. We want to include
+ * all external URLs even if they have dangerous protocols.
+ *
+ * @param string $url
+ *
+ * @return bool
+ *   TRUE if URL has protocol.
+ */
+function _filter_url_has_protocol($url) {
+  // If it has a scheme then it has a protocol.
+  if (parse_url($url, PHP_URL_SCHEME)) {
+    return TRUE;
+  }
+  return FALSE;
+}
+
+/**
+ * Given a URL, find an alias and matching lanaguage.
+ *
+ * @param $url
+ *   The path to investigate for corresponding aliases.
+ *
+ * @return object|FALSE
+ *   The langcode.
+ */
+function _filter_lookup_locale_path($url) {
+  $languages = language_list(TRUE);
+    foreach ($languages as $language) {
```

Comment:
Shouldn't be indented...

But my actual question is: what about '/de/node/5'? Does this handle language prefix plus node/nid corrently?

Reply:
Removed function.